RFC: Partitioned proofs (#4228)
model-checking/kani: rfc/src/rfcs/0014-harness-partition.md

    assert!(target_fn(input) > 0)
}
```

However, this can affect soundess, as there's no guarantee that your partitions will fully span the space of possible inputs. The only way to determine that a set of proofs like the one above are incorrect (as it forgets to account for when `i == 0`) is by manual inspection, which gets infeasible for proofs with complex types or partition rules.

It would be helpful if Kani provided this as a built-in feature that could reason about given partition conditions to provide soundess guarantees.

## User Experience

The current thought is for users to interact with this using a new `#[kani::partitioned_proof()]` attribute where they provide the conditions by which to partition the input space of their proof. Each condition must be of type `fn(&T) -> bool` (where `T` is the input type to the proof body that implements `kani::Arbitrary`) and is used to filter which values are part of that partition.

For example, the above would become

```rust
#[kani::partitioned_proof(|a: &i32| { *a > 0 }, |a: &i32| { *a < 0 })]
pub fn partitioned_harness(input: i32) {
    assert!(target_fn(input) > 0)
}
```

> And Kani would automatically handle checking the partition conditions for soundess and generating the partitioned proofs. Overlaps would be allowed as they don't affect soundess and may be useful in certain cases (see [below](#2-checking-for-overlapping-partitions) for more discussion).

Author: The thought (and implementation in the previous prototype) is that you would generate a new proof harness that declares `input` as a `kani::any::<i32>()` and asserts that at least one of the partition conditions evaluates to true. This, conceptually, should ensure that any `i32` value will fall into at least one partition.

## Open questions

1. What's the best way to ensure that type errors in the `kani::partitioned_proof` macro are clear, understandable and actionable for users?
2. Can function names be generated more prettily while still remainging unique? (the current hashes are often opaque for users)
3. Are there any additional correctness issues introduced by this approach?
4. Is there a way to simplify the generated `..._full_coverage()` proof harness? The iteration seems to be instrumented by our code which adds additional checks.
5. Is there a way to promote the fact that verification failed on the `..._full_coverage()` proof harness as an error that could affect soundess? If Kani's users tend to look into any failed checks, this may not be needed, but it may be possible for a single failure to get lost in big projects.

> 6. Would it be desirable to add an optional argument that specifies which subset of inputs you're trying to verify with the proof as a whole (if not provided just the whole input space)?

Author: Do you think it makes sense to make users still provide a filter that they wish to cover, even in the `unsafe_partition` case? For most use cases I can imagine (including the 64-bit int multiplication), it seems like you'd still have a well-defined range of inputs you want to verify rather than entirely avoiding a coverage check. Although the latter definitely provides the user more freedom.

> 6. Would it be desirable to add an optional argument that specifies which subset of inputs you're trying to verify with the proof as a whole (if not provided just the whole input space)?

Reviewer: > For most use cases I can imagine (including the 64-bit int multiplication), it seems like you'd still have a well-defined range of inputs you want to verify rather than entirely avoiding a coverage check
Yeah, I think this is right. For the scope of this RFC, though, I would leave it as an open question; I don't think we should implement any kind of unsafe partition until the safe one is merged as an unstable feature. Once it is, we can open a tracking issue for open questions that links back to the RFC.

```
Checking partition #1 (*input > 0) of harness partitioned_proof...
CBMC 6.7.1 (cbmc-6.7.1)
/* MORE CBMC OUTPUT */
RESULTS:
Check 1: partitioned_proof_partition_1.assertion.1
	 - Status: SUCCESS
	 - Description: "assertion failed: target_fn(input) > 0"
	 - Location: src/pre_partition.rs:8:5 in function partitioned_proof_partition_1
     - Partition: #1 - *input > 0

Checking partition #2 (*input < 0) of harness partitioned_proof...
CBMC 6.7.1 (cbmc-6.7.1)
/* MORE CBMC OUTPUT */
RESULTS:
Check 1: partitioned_proof_partition_2.assertion.1
	 - Status: SUCCESS
	 - Description: "assertion failed: target_fn(input) > 0"
	 - Location: src/pre_partition.rs:8:5 in function partitioned_proof_partition_2
     - Partition: #2 - *input < 0

Checking coverage of partitions for harness partitioned_proof...
CBMC 6.7.1 (cbmc-6.7.1)
/* MORE CBMC OUTPUT */
RESULTS:
Check 1: partitioned_proof_coverage.assertion.1
	 - Status: FAILURE
	 - Description: "partitions for partitioned_proof do not cover all possible values of the i32 type."
	 - Location: src/pre_partition.rs:11:5 in function partitioned_proof_coverage
```

> 

Reviewer 2: We could run the coverage harness with concrete playback to generate a test.

> 

Reviewer: Nice idea! @AlexanderPortland let's add this to the RFC as an open question and put this coverage check as a potential solution... don't think this is necessary for a first implementation but a nice-to-have for sure.

This would be along the lines of:
```rust
pub fn partitioned_proof_coverage() {
    /* OTHER CODE BEFORE THE CALL TO `kani::partition` IF THERE WAS ANY */
    let partition_variable = kani::any();

    // checks that any possible value of partition_variable will fall into at least one partition.
    let is_in_partitions = [|a: &i32| *a > 0, |a: &i32| *a < 0].into_iter().any(|cond| cond(&partition_variable));
    kani::assert(is_in_partitions, "partition conditions for partitioned_proof do not cover all possible values of a i32.");
}
```

### Corner cases
Although the fact that the partition conditions are function pointers prevents them from capturing any local variables, their value can still be affected by non-deterministic values.

```rust
#[kani::proof]
pub fn strange_partitioned_proof() {
    let f: fn(&i32) -> bool = if kani::any::<i32>() > 10 {
        |a: &i32| *a < 0
    } else {
        |a: &i32| *a < -2
    };

    kani::partition([f, |a: &i32| *a < 0], ...);
}
```

The behavior of a partition in this case seems ill-defined, so any call to `kani::partition` with conditions that are not constant closures would be disallowed.

> ## Rationale and alternatives

Reviewer: I would expand this section a bit with some rationale for your design, namely:

- Why do we have one round of CBMC output per partitioned harness? (One could imagine a design where you combine the results back together in postprocessing so you can just display "Checking harness..."). I think the splitting is 1) easier to implement for a first draft and 2) perhaps easier to debug but has the con of being longer.
- Why is `kani::partition` a function, rather than a proof attribute or a different design?

You've done a lot of thinking to settle on this design, so it'd be good to reflect that in the writing by (briefly) mentioning other points in the design space and why you decided they weren't as good.

### 1. Allowing overlapping partition conditions

> As a corollary to checking that partitions span the space of all possible inputs, I had initially considered giving a warning if partitions are overlapping as this could indicate ill-defined bounds.

Reviewer: I feel like a warning could be interesting future work but I agree it's not necessary for the feature to be useful

However, this kind of overlap would not affect the correctness of the partition, as every possible value of the partition variable would still be checked, just potentially more than once. Admittedly, it's fairly difficult to come up with a realistic use case in which an overlap is helpful for users, so whether this should be allowed is up for debate.

## Open questions

1. How can we ensure that type errors in the `kani::partition` function call are clear, understandable and actionable for users?
2. How can we test that there are no correctness issues introduced by our partition implementation?

> 3. Should verification fully panic if a partitioned proof is lacking full coverage or just have that one assertion fail verification?

Reviewer: Not sure I understand what you mean here--Kani certainly shouldn't ICE but I presume that's not your point? If at least one property fails, verification will fail.

Author: My question was more along the lines of should we not even give a verification output and have a pretty internal kani error saying something like "your partition bounds are incomplete, so any output we could give you for this proof would be incomplete -> go fix that". Maybe a well-named property failing is already enough though

Reviewer: I think compiler error is always better than verification error but not sure atm how you'd implement it on the compiler side. So not a huge deal.

4. What's the best way to provide support for partitioning `BoundedArbitrary` types?
Potentially through a separate function `partition_bounded` which would be generic over `T: BoundedArbitrary` and a `const N: usize` which would be used for the call to `kani::bounded_any()`.

> 5. How should users be able to control our coverage checks? Does it make sense to provide another function that makes the user specify a smaller domain that we can still prove coverage for (or skips those checks entirely)?

Reviewer: ```suggestion
5. How should users be able to control our coverage checks?
Users may only want to test a subset of the input space (e.g., only testing multiplication for certain integer ranges), but the current design mandates that they cover the whole input space.
We could potentially introduce a separate API, e.g. `unsafe_partition`, that allows this, or some other mechanism to skip the coverage check.
```

- **Feature Name:** Partitioned Proofs (`partition-proof`)
- **Feature Request Issue:** [#3006](https://github.com/model-checking/kani/issues/3006)
- **RFC PR:** https://github.com/model-checking/kani/pull/4228
- **Status:** Under Review
- **Version:** 0
- **Proof-of-concept:** None yet, the previous prototype uses an out of date API.

-------------------

## Summary

> It can often be useful to subdivide an expensive proof harness so that different parts of the input space are verified separately.

Reviewer 2: Worth pointing out that this is also known as proof by cases or case split

Adding the built-in ability to partition a proof harness into different pieces (that each make differing assumptions about their inputs) could reduce the cost of expensive proofs, while allowing Kani to automatically check that the partitions cover the entire input space and, thus, will be equivalent to a single-harness proof.

## User Impact

Imagine that you have a function to verify like the following (based on the example from [#3006](https://github.com/model-checking/kani/issues/3006)).

```rust
pub fn target_fn(input: i32) -> isize {
    let val = if input > 0 {
        very_complex_fn_1(input)
    } else {
        very_complex_fn_2(input)
    };
    very_complex_fn_3(val)
}

#[kani::proof]
pub fn proof_harness() {
    let input = kani::any();
    assert!(target_fn(input) > 0)
}
```

Since there are two tricky to analyze function calls, but only one will ever be called on a given input, you might want to verify all values of `input` where `input > 0` that will take the first branch separately from those that will take the second.

> This way, each solve will only have to reason about two of the three complex function calls, and you could use Kani's parallel proof runner to run both proofs at once.

Reviewer 2: ```suggestion
This way, each verification run will only have to reason about two of the three complex function calls, and you could use Kani's parallel proof runner to run both proofs at once.
```

The best way to currently do this is by manually partitioning out these paths into two proof harnesses.

```rust
#[kani::proof]
pub fn first_branch_harness() {
    let input = kani::any_where(|i: &i32| *i > 0i32);
    assert!(target_fn(input) > 0)
}

#[kani::proof]
pub fn second_branch_harness() {
    let input = kani::any_where(|i: &i32| *i < 0i32); // ERROR: This should've been i <= 0
    assert!(target_fn(input) > 0)
}
```

One could also write preconditions for `very_complex_fn_{1_2}` that restrict `input` accordingly, write `proof_for_contract` harnesses for those, and then use `stub_verified` on the `target_fn` harness. 

However, either of these strategies:
- **can affect soundness**--there's no guarantee that your partitions will fully span the space of possible inputs.

> The only way to determine that a set of proofs like the one above are incorrect (as it forgets to verify the value of 0 for `input`) is by manual inspection.

Reviewer 2: ```suggestion
The only way to determine that a set of proofs like the one above are incomplete (as it forgets to verify the value of 0 for `input`) is by manual inspection.
```

This gets infeasible for proofs with complex partition rules like those found in the [proofs for the standard library's unchecked multiplication](https://github.com/model-checking/verify-rust-std/blob/1c4ea17a99b9202f96608473083998b116bb6508/library/core/src/num/mod.rs#L1818-L1836).
- **increases user burden**--instead of having to write and maintain a single proof, the user now has to handle a proof for each partition.

Instead, Kani should provide a feature to automatically partition a proof based on certain conditions and ensure that the newly partitioned harnesses are equivalent to a single-harness proof.

## User Experience

Kani will introduce a new `kani::partition` function, gated behind an unstable feature flag.
When calling this function, the user would provide the type of the partition variable, as well as a set of partition conditions and a closure with the actual code to run on the partitioned variable (we'll call this the *partition closure*).

For example, the above would become

```rust
#[kani::proof]
pub fn partitioned_proof() {

>     kani::partition::<i32>([|a| *a > 0, |a| *a < 0], |input: i32| target_fn(input));

Reviewer 2: Should the assertion be added  make the harness complete?
```rust
    assert!(target_fn(input) > 0)
```

>     kani::partition::<i32>([|a| *a > 0, |a| *a < 0], |input: i32| target_fn(input));

Reviewer 2: Also, don't we need to specify the other generics?
```suggestion
    kani::partition::<i32, _, _>([|a| *a > 0, |a| *a < 0], |input: i32| target_fn(input));
```

Checking partition #2 (*input < 0) of harness partitioned_proof...
CBMC 6.7.1 (cbmc-6.7.1)
/* MORE CBMC OUTPUT */
RESULTS:
Check 1: partitioned_proof_partition_2.assertion.1
     - Status: SUCCESS
     - Description: "assertion failed: target_fn(input) > 0"
     - Location: src/pre_partition.rs:8:5 in function partitioned_proof_partition_2

Checking coverage of partitions for harness partitioned_proof...
CBMC 6.7.1 (cbmc-6.7.1)
/* MORE CBMC OUTPUT */
RESULTS:
Check 1: partitioned_proof_coverage.assertion.1
     - Status: FAILURE
     - Description: "partitions for partitioned_proof do not cover all possible values of the i32 type."
     - Location: src/pre_partition.rs:11:5 in function partitioned_proof_coverage
```

Note that the coverage check fails, since we're missing the `i = 0` case.

## Software Design

This change would introduce the new `kani::partition` function, with the following signature:

```rust
pub fn partition<T: Arbitrary, R, const N: usize>(
    conditions: [fn(&T) -> bool; N],
    and_run: fn(T) -> R, // the "partition closure"
) -> R
```

This signature makes a few key design decisions:

> - *It takes in an array of `conditions` with the specific length `const N`*, ensuring the number of harnesses we have to generate is knowable at compile time when we're doing partition generation.

Reviewer 2: ```suggestion
- *It takes in an array of `conditions` with the specific length `const N`*, ensuring the number of harnesses we have to generate is known at compile time when we're doing partition generation.
```

- *Partition conditions & the partition closure are represented by function pointers*.
This allows the user to provide closures of the same signature, but only if those closures do not capture any state.
Allowing conditions closures to capture runtime state like local variables would be unwise as partitions are conceptually static divisions of a proof which should not depend on runtime values.
- *The function returns the same type `R` as the underlying partition closure*, with the thought being that value could then be used in later computation without affecting the mechanics of a partition.
- *The partition variable is bound by `T: Arbitrary`*.
This would allow us to construct a new non-deterministic value with `kani::any()` that can then be passed to the partition closure.

> When the `kani-compiler` encounters a call to this function, it will transparently replace the entire `#[kani::proof()]` that called it with a set of new generated functions that, as a whole, will represent the partitioned proof. This new partitioned proof would contain:

Reviewer 2: ```suggestion
When the `kani-compiler` encounters a call to this function, it will transparently replace the entire `#[kani::proof]` that called it with a set of new generated functions that, as a whole, will represent the partitioned proof. This new partitioned proof would contain:
```

1. one new proof harness for each partition

In the above example, this would be along the lines of:

```rust
#[kani::proof]
/// (auto-generated partition of `input` for condition `|a| *a > 0`)
pub fn partitioned_proof_partition_1() {
    /* OTHER CODE BEFORE THE CALL TO `kani::partition` IF THERE WAS ANY */
    let partition_variable = kani::any::<i32>();

>     kani::assume((|a: &i32| *a > 0)(&partition_variable));

Reviewer 2: Neat!
